### PR TITLE
Fix Go Report Card Issues

### PR DESCRIPTION
**Pull request by HakanVardarr**
I fixed all of the issues listed on goreportcard.
Screenshot 2024-03-30 at 09 52 41

### CODE REVIEW

**Review by ccoVeille — CHANGES_REQUESTED** [review]
As I said in https://github.com/Boeing/config-file-validator/issues/134#issuecomment-2028109647 you fixed #134 issue.

But I would recommend considering simplifying your code like this

```go
	// Group the output if the user specified a group by option
	// Length is equal to one when empty as it contains an empty string
	if len(GroupOutput) == 1 && GroupOutput[0] != "" {
		// Check reporter type to determine how to print
		if _, ok := c.Reporter.(reporter.JsonReporter); ok {
			return reporter.PrintSingleGroupJson(reportGroup.(map[string][]reporter.Report))
		}
		return reporter.PrintSingleGroupStdout(reportGroup.(map[string][]reporter.Report))
	} else if len(GroupOutput) == 2 {
		if _, ok := c.Reporter.(reporter.JsonReporter); ok {
			return reporter.PrintDoubleGroupJson(reportGroup.(map[string]map[string][]reporter.Report))
		}
		return reporter.PrintDoubleGroupStdout(reportGroup.(map[string]map[string][]reporter.Report))
	} else if len(GroupOutput) == 3 {
		if _, ok := c.Reporter.(reporter.JsonReporter); ok {
			return reporter.PrintTripleGroupJson(reportGroup.(map[string]map[string]map[string][]reporter.Report))
		}
		return reporter.PrintTripleGroupStdout(reportGroup.(map[string]map[string]map[string][]reporter.Report))
	}
	return c.Reporter.Print(reports)
```

**Review by HakanVardarr — COMMENTED** [review]
I added these refactorings to the PR thanks.

**Review by ccoVeille — CHANGES_REQUESTED** [review]
because you are now returning different type result, you are using `interface{}` here

It leads you to play with type assertion when calling the reporters.

I would suggest you to write something like this

```go

// printReports prints the reports based on the specified grouping and reporter type.
// It returns any error encountered during the printing process.
func (c CLI) printReports(reports []reporter.Report) error {
	if Quiet {
		return nil
	}

	switch len(GroupOutput) {
	case 1:
		return printGroupSingle(reports)
	case 2:
		return printGroupDouble(reports)
	case 3:
		return printGroupTriple(reports)
	default:
		return nil, fmt.Errorf("Invalid number of group outputs: %d", len(GroupOutput))
	}
}

func (c CLI) printGroupSingle(reports []reporter.Report) error {
	if GroupOutput[0] == "" {
		return c.Reporter.Print(reports)
	}

	reportGroup, err := c.GroupBySingle(reports, GroupOutput[0])
	if err != nil {
		return err
	}

	// Check reporter type to determine how to print
	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
		return reporter.PrintSingleGroupJson(reportGroup)
	}

	return reporter.PrintSingleGroupStdout(reportGroup)
}

func (c CLI) printGroupDouble(reports []reporter.Report) error {
	reportGroup, err := c.GroupByDouble(reports, GroupOutput)
	if err != nil {
		return err
	}

	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
		return reporter.PrintDoubleGroupJson(reportGroup)
	}

	return reporter.PrintDoubleGroupStdout(reportGroup)
}

func (c CLI) printGroupTriple(reports []reporter.Report) error {
	reportGroup, err := c.GroupByTriple(reports, GroupOutput)
	if err != nil {
		return err
	}

	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
		return reporter.PrintTripleGroupJson(reportGroup)
	}
	return reporter.PrintTripleGroupStdout(reportGroup)
}

```

This is mostly pseudocode made in a web browser, it may need some adjustments

**Comment by ccoVeille** [comment]
> I added these refactorings to the PR thanks.

May I ask a `Co-Authored-By` tag in your commit message then ? 😆

https://docs.github.com/articles/creating-a-commit-with-multiple-authors#:~:text=a%20remote%20repository.%22-,Creating%20co%2Dauthored%20commits%20on%20GitHub,address%20for%20each%20co%2Dauthor.

**Comment by HakanVardarr** [comment]
> > I added these refactorings to the PR thanks.
> 
> May I ask a `Co-Authored-By` tag in your commit message then ? 😆
> 
> https://docs.github.com/articles/creating-a-commit-with-multiple-authors#:~:text=a%20remote%20repository.%22-,Creating%20co%2Dauthored%20commits%20on%20GitHub,address%20for%20each%20co%2Dauthor.
> 
> Mine would be something like
> 
> ```
> Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
> ```

Is this right? First time doing this. :)

**Comment by ccoVeille** [comment]
> > > I added these refactorings to the PR thanks.
> > 
> > 
> > May I ask a `Co-Authored-By` tag in your commit message then ? 😆
> > https://docs.github.com/articles/creating-a-commit-with-multiple-authors#:~:text=a%20remote%20repository.%22-,Creating%20co%2Dauthored%20commits%20on%20GitHub,address%20for%20each%20co%2Dauthor.
> > Mine would be something like
> > ```
> > Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
> > ```
> 
> Is this right? First time doing this. :)

I think you should have a new line between the commit message and the `Co-authored-by:`

**Comment by ccoVeille** [comment]
> because you are now returning different type result, you are using `interface{}` here
> 
> It leads you to play with type assertion when calling the reporters.
> 
> I would suggest you to write something like this
> 
> ```go
> // printReports prints the reports based on the specified grouping and reporter type.
> // It returns any error encountered during the printing process.
> func (c CLI) printReports(reports []reporter.Report) error {
> 	if Quiet {
> 		return nil
> 	}
> 
> 	switch len(GroupOutput) {
> 	case 1:
> 		return printGroupSingle(reports)
> 	case 2:
> 		return printGroupDouble(reports)
> 	case 3:
> 		return printGroupTriple(reports)
> 	default:
> 		return nil, fmt.Errorf("Invalid number of group outputs: %d", len(GroupOutput))
> 	}
> }
> 
> func (c CLI) printGroupSingle(reports []reporter.Report) error {
> 	if GroupOutput[0] == "" {
> 		return c.Reporter.Print(reports)
> 	}
> 
> 	reportGroup, err := c.GroupBySingle(reports, GroupOutput[0])
> 	if err != nil {
> 		return err
> 	}
> 
> 	// Check reporter type to determine how to print
> 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> 		return reporter.PrintSingleGroupJson(reportGroup)
> 	}
> 
> 	return reporter.PrintSingleGroupStdout(reportGroup)
> }
> 
> func (c CLI) printGroupDouble(reports []reporter.Report) error {
> 	reportGroup, err := c.GroupByDouble(reports, GroupOutput)
> 	if err != nil {
> 		return err
> 	}
> 
> 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> 		return reporter.PrintDoubleGroupJson(reportGroup)
> 	}
> 
> 	return reporter.PrintDoubleGroupStdout(reportGroup)
> }
> 
> func (c CLI) printGroupTriple(reports []reporter.Report) error {
> 	reportGroup, err := c.GroupByTriple(reports, GroupOutput)
> 	if err != nil {
> 		return err
> 	}
> 
> 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> 		return reporter.PrintTripleGroupJson(reportGroup)
> 	}
> 	return reporter.PrintTripleGroupStdout(reportGroup)
> }
> ```
> 
> This is mostly pseudocode made in a web browser, it may need some adjustments

did you see this @HakanVardarr BTW? what are your thoughts about it ?

**Comment by HakanVardarr** [comment]
> > because you are now returning different type result, you are using `interface{}` here
> > It leads you to play with type assertion when calling the reporters.
> > I would suggest you to write something like this
> > ```go
> > // printReports prints the reports based on the specified grouping and reporter type.
> > // It returns any error encountered during the printing process.
> > func (c CLI) printReports(reports []reporter.Report) error {
> > 	if Quiet {
> > 		return nil
> > 	}
> > 
> > 	switch len(GroupOutput) {
> > 	case 1:
> > 		return printGroupSingle(reports)
> > 	case 2:
> > 		return printGroupDouble(reports)
> > 	case 3:
> > 		return printGroupTriple(reports)
> > 	default:
> > 		return nil, fmt.Errorf("Invalid number of group outputs: %d", len(GroupOutput))
> > 	}
> > }
> > 
> > func (c CLI) printGroupSingle(reports []reporter.Report) error {
> > 	if GroupOutput[0] == "" {
> > 		return c.Reporter.Print(reports)
> > 	}
> > 
> > 	reportGroup, err := c.GroupBySingle(reports, GroupOutput[0])
> > 	if err != nil {
> > 		return err
> > 	}
> > 
> > 	// Check reporter type to determine how to print
> > 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> > 		return reporter.PrintSingleGroupJson(reportGroup)
> > 	}
> > 
> > 	return reporter.PrintSingleGroupStdout(reportGroup)
> > }
> > 
> > func (c CLI) printGroupDouble(reports []reporter.Report) error {
> > 	reportGroup, err := c.GroupByDouble(reports, GroupOutput)
> > 	if err != nil {
> > 		return err
> > 	}
> > 
> > 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> > 		return reporter.PrintDoubleGroupJson(reportGroup)
> > 	}
> > 
> > 	return reporter.PrintDoubleGroupStdout(reportGroup)
> > }
> > 
> > func (c CLI) printGroupTriple(reports []reporter.Report) error {
> > 	reportGroup, err := c.GroupByTriple(reports, GroupOutput)
> > 	if err != nil {
> > 		return err
> > 	}
> > 
> > 	if _, ok := c.Reporter.(reporter.JsonReporter); ok {
> > 		return reporter.PrintTripleGroupJson(reportGroup)
> > 	}
> > 	return reporter.PrintTripleGroupStdout(reportGroup)
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is mostly pseudocode made in a web browser, it may need some adjustments
> 
> did you see this @HakanVardarr BTW? what are your thoughts about it ?

Oh I didn't see these, I'm looking right now

**Comment by HakanVardarr** [comment]
I think that looks good. I'll add it and co-auther you.

**Comment by HakanVardarr** [comment]
I added the co-authered-by part @ccoVeille is it right?

**Comment by ccoVeille** [comment]
It's ok 305ebbabb5c76d9aafe50cc4df1efb970a8591bc

You can see here our names in 
image

You can also see it here 

image

The right pattern to use is:

```
Commit message

Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
```

And if you have to add a description it would be

```
Commit message

Description line1
Description line2
(...)

Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
```

File: pkg/finder/finder_test.go [hunk 1]
```diff
@@ -2,10 +2,11 @@ package finder
 
 import (
 	"fmt"
-	"github.com/Boeing/config-file-validator/pkg/misc"
 	"path/filepath"
 	"testing"
 
+	"github.com/Boeing/config-file-validator/pkg/misc"
+
```
Comment (ccoVeille) [inline pkg/finder/finder_test.go:9]: import should be grouped altogether, so it's ok but you shouldn't have this empty line

**Comment by ccoVeille** [comment]
Thanks for following my suggestion. I think we are good.

The code is now clearer and easier to maintain.

**Comment by ccoVeille** [comment]
You just have to check if all tests are good.

I'm a bit worried that your version will report nothing in json mode if quiet is enabled.

So maybe, you will have to remove the if quiet and return nil in each print single/double/triple when quiet is true and it's not about dumping json

**Comment by HakanVardarr** [comment]
> You just have to check if all tests are good.
> 
> I'm a bit worried that your version will report nothing in json mode if quiet is enabled.
> 
> So maybe, you will have to remove the if quiet and return nil in each print single/double/triple when quiet is true and it's not about dumping json

I think this commit solves the issue.

**Review by kehoecj — COMMENTED** [review]
Thank you for the PR @HakanVardarr ! Looks like some good improvements. Can you take a look at the failing unit tests?

**Comment by HakanVardarr** [comment]
I hope they are right I got midterm tomorrow I don't have much time to debug :) @kehoecj

**Comment by kehoecj** [comment]
> I hope they are right I got midterm tomorrow I don't have much time to debug :) @kehoecj 

No rush! Midterm is more important 😀

**Comment by HakanVardarr** [comment]
> > I hope they are right I got midterm tomorrow I don't have much time to debug :) @kehoecj
> 
> No rush! Midterm is more important 😀

I handled the unit tests but coverage is stuck at 94.9% 😄

**Comment by ccoVeille** [comment]
This may help you

```shell
go test -coverprofile=cover.out ./... ; go tool cover -html=cover.out -o cover.html
```

You will see you can improve coverage in

```go
      if *reportTypePtr == "junit" && *groupOutputPtr != "" {
                fmt.Println("Wrong parameter value for reporter, groupby is not supported for JUnit reports")
                flag.Usage()
                return validatorConfig{}, errors.New("Wrong parameter value for reporter, groupby is not supported for JUnit reports")
        }
```

and

```go
                        if _, ok := seenValues[groupBy]; ok {
                                fmt.Println("Wrong parameter value for groupby, duplicate values are not allowed")
                                flag.Usage()
                                return validatorConfig{}, errors.New("Wrong parameter value for groupby, duplicate values are not allowed")
                        }
```

**Comment by HakanVardarr** [comment]
> This may help you
> 
> ```shell
> go test -coverprofile=cover.out ./... ; go tool cover -html=cover.out -o cover.html
> ```
> 
> You will see you can improve coverage in
> 
> ```go
>       if *reportTypePtr == "junit" && *groupOutputPtr != "" {
>                 fmt.Println("Wrong parameter value for reporter, groupby is not supported for JUnit reports")
>                 flag.Usage()
>                 return validatorConfig{}, errors.New("Wrong parameter value for reporter, groupby is not supported for JUnit reports")
>         }
> ```
> 
> and
> 
> ```go
>                         if _, ok := seenValues[groupBy]; ok {
>                                 fmt.Println("Wrong parameter value for groupby, duplicate values are not allowed")
>                                 flag.Usage()
>                                 return validatorConfig{}, errors.New("Wrong parameter value for groupby, duplicate values are not allowed")
>                         }
> ```

Yep they worked perfect. Thanks, I Co-Authered you to the commit.

**Comment by ccoVeille** [comment]
I would say the number of commits could be reduced by grouping them with an interactive rebase. But I will let maintainers tell what they want.

**Comment by HakanVardarr** [comment]
> I would say the number of commits could be reduced by grouping them with an interactive rebase. But I will let maintainers tell what they want.

I don't know much about git stuffs. I am new to this that is why commit number is high.

**Comment by ccoVeille** [comment]
> > I would say the number of commits could be reduced by grouping them with an interactive rebase. But I will let maintainers tell what they want.
> 
> I don't know much about git stuffs. I am new to this that is why commit number is high. 

It's OK. Let's wait for the maintainers point of view. Have fun with your midterms. That's more important to worry about.

**Comment by HakanVardarr** [comment]
> > > I would say the number of commits could be reduced by grouping them with an interactive rebase. But I will let maintainers tell what they want.
> > 
> > I don't know much about git stuffs. I am new to this that is why commit number is high. 
> 
> It's OK. Let's wait for the maintainers point of view. Have fun with your midterms. That's more important to worry about.

I'll try to have fun. 😅

**Review by kehoecj — APPROVED** [review]
LGTM - thanks @HakanVardarr fixing the report card issues and general cleanup

**Comment by ccoVeille** [comment]
Great @HakanVardarr, I can rebase mine now.

I hope you enjoyed my help on your PR.

**Comment by ccoVeille** [comment]
@kehoecj I appreciate you squashed all the changes: 9006fce

I was unsure the Co-Authored-By would be kept, but I'm happy to see it worked

**Comment by HakanVardarr** [comment]
> Great @HakanVardarr, I can rebase mine now.
> 
> I hope you enjoyed my help on your PR.

I enjoyed your help. Thanks <3

**Pull request merged**